File: download_manifest_mods.py
import os
from shutil import move
import json
from time import sleep
import requests

from download_file import download


def download_manifest_mods(path,apiKey):
    with open(path, encoding="UTF-8") as f:
        data = json.load(f)
        print("Starting download of modpack manifest server mods...")
        for mod in data['files']:
            try:
                mod_id = mod['projectID']
                file_id = mod['fileID']
            except:
                print("Error parsing mod download IDs")
            if mod['required'] == False:
                print(f"Detected optional server mod {mod_id}. Skipping download.")
                continue
            if mod_id and file_id:
                # Try and get Mod Download file from our CF API
                try:
                    print(f"https://api.hypesrv.net/v2/modpack/fileUrl/{mod_id}/{file_id}")
                    headers = {"Authorization": "Bearer "+apiKey}
                    print(headers)
                    response_body = requests.get(f"https://api.hypesrv.net/v2/modpack/fileUrl/{mod_id}/{file_id}", timeout=60, headers=headers).json()
                    print(response_body)
                    response = response_body["data"]
                    mname = response["name"]
                    print(f"Downloading {mname} ...")
                    download(response["download_url"])
                except:
                    print(f"Error downloading mod {mod_id}.")

        sleep(1)
        print("Finished downloading all server mods from modpack manifest.")

File: src/downloadManifestMods.ts
import fs from 'fs-extra';
import axios from 'axios';
import { download } from './downloadFile.js';
import { log } from './util/logger.js';

interface ManifestModEntry {
    projectID?: number | string;
    projectId?: number | string;
    fileID?: number | string;
    fileId?: number | string;
}

interface Manifest {
    files?: ManifestModEntry[];
}

interface HypesrvFileInfo {
    downloadUrl?: string;
    displayName?: string;
    fileName?: string;
}

interface HypesrvResponse {
    data?: HypesrvFileInfo;
}

export async function downloadManifestMods(
    manifestPath: string,
    apiKey: string | false
): Promise<void> {
    const raw = await fs.readFile(manifestPath, 'utf8');
    const data = JSON.parse(raw) as Manifest;

    log.info('Starting download of manifest server mods...');
    for (const mod of data.files || []) {
        const modId = mod.projectID ?? mod.projectId;
        const fileId = mod.fileID ?? mod.fileId;
        if (!modId || !fileId) continue;

        try {
            const url = `https://api.hypeserv.net/v3/modpack/fileUrl/${modId}/${fileId}`;
            log.info(url);
            const headers: Record<string, string> = apiKey
                ? { Authorization: `Bearer ${apiKey}` }
                : {};
            const resp = await axios.get<HypesrvResponse>(url, {
                timeout: 60000,
                headers
            });
            const info = resp.data?.data;
            if (!info?.downloadUrl) {
                log.warn(`No downloadUrl for mod ${modId}/${fileId}`);
                continue;
            }
            log.info(`Downloading ${info.displayName || info.fileName}...`);
            await download(info.downloadUrl);
        } catch (err) {
            const message = err instanceof Error ? err.message : String(err);
            log.warn(`Error downloading mod ${modId}/${fileId}: ${message}`);
        }
    }
    log.info('Finished downloading all server mods from modpack manifest.');
}
